### PR TITLE
[Performance] Faster `CatFrames.unfolding` with `padding="same"`

### LINKED ISSUE #2406: [BUG] CatFrames with `padding="same"` is slow
## Describe the bug

A discord user reported that `CatFrames` is slow and gave a repro script: <https://discord.com/channels/1171857748607115354/1268244652180377732/1270523213600002129>

## To Reproduce

The given repro script allocated more memory than my computer can handle, so I modified it like so:

```python
import torch
from torchrl.envs import CatFrames, Compose
from torchrl.data import PrioritizedSliceSampler, ReplayBuffer, LazyTensorStorage
from tensordict import TensorDict
import timeit

device = 'cpu'

def catframes_speed_test(with_catframes=True):
    if with_catframes:
        obs_shape = (4, 42, 42, 1)
        transform = Compose(
            CatFrames(N=4, dim=-1, in_keys=["observation"], done_key="done"),
            CatFrames(N=4, dim=-1, in_keys=[("next", "observation")], done_key="done"))
    else:
        obs_shape = (4, 42, 42, 4)
        transform = None


    max_size = 100_000

    sampler = PrioritizedSliceSampler(
        max_capacity=max_size,
        alpha=0.5,
        beta=0.4,
        strict_length=False,
        num_slices=32 // 4,
        span=[True, False],
    )

    exp_buffer = ReplayBuffer(
        storage=LazyTensorStorage(max_size=max_size, device=device),
        sampler=sampler,
        batch_size=32,
        transform=transform,
    )

    fake_data = TensorDict(
        {
            "observation": torch.zeros(obs_shape, dtype=torch.float32, device=device),
            "next": {"observation": torch.zeros(obs_shape, dtype=torch.float32, device=device),
                     "done": torch.zeros((obs_shape[0], 1), dtype=torch.bool, device=device)},
        },
        batch_size=[obs_shape[0]],
        device=device,
    )

    for _ in range(25):
        exp_buffer.extend(fake_data)

    num_iters = 30
    timer = timeit.Timer(
        stmt=lambda: exp_buffer.sample(return_info=True),
    )
    time_per_iter = timer.timeit(num_iters) / num_iters
    print(f"Sampling took {time_per_iter} seconds.")

    exp_buffer.empty()


if __name__ == "__main__":
    print("WITHOUT CATFRAMES")
    catframes_speed_test(with_catframes=False)

    print("WITH CATFRAMES")
    catframes_speed_test(with_catframes=True)
```

On my machine, I get this output:

```
WITHOUT CATFRAMES
Sampling took 0.0016307519336502688 seconds.
WITH CATFRAMES
Sampling took 0.23012657026653566 seconds.
```

## Expected behavior

Adding the CatFrames transformation to a replay buffer should not slow it down significantly, but in the above case, it caused a slowdown of more than 100x.

## Checklist

- [x] I have checked that there is no similar issue in the repo (**required**)
- [x] I have read the [documentation](https://github.com/pytorch/rl/tree/main/docs/) (**required**)
- [x] I have provided a minimal working example to reproduce the bug (**required**)

### PR DESCRIPTION
**Pull request by kurtamohler**
## Description

Improves the performance of `CatFrames.unfolding` with `padding="same"`. Running the performance measurement script in the issue description of #2406 shows a speedup of ~70x compared to the previous implementation:

```
WITHOUT CATFRAMES
Sampling took 0.0016054896002363724 seconds.
WITH CATFRAMES
Sampling took 0.003015774066928619 seconds.
```

## Motivation and Context

close #2406

- [x] I have raised an issue to propose this change ([required](https://github.com/pytorch/rl/issues) for new features and bug fixes)

## Types of changes

What types of changes does your code introduce? Remove all that do not apply:

- [x] Bug fix (non-breaking change which fixes an issue)

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [x] I have read the [CONTRIBUTION](https://github.com/pytorch/rl/blob/main/CONTRIBUTING.md) guide (**required**)
- [ ] My change requires a change to the documentation.
- [ ] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Review by vmoens — APPROVED** [review]
Amazing, I'm so grateful for this!
Can you just comment on the 0: indexing?

File: torchrl/envs/transforms/transforms.py [hunk 1]
```diff
@@ -3082,6 +3082,33 @@ def forward(self, tensordict: TensorDictBase) -> TensorDictBase:
...
+
+        if d - 1 > extra_dims:
+            res_flat_series_flat_batch = res_flat_series.flatten(1, d - 1)
+        else:
+            res_flat_series_flat_batch = res_flat_series[:, None]
+
+        for sample_idx, num_repeats in enumerate(num_repeats_per_sample):
+            if num_repeats > 0:
+                res_slice = res_flat_series_flat_batch[sample_idx]
+                res_slice[:, 0:num_repeats] = res_slice[
```
[the remark this answers is not in the document]
Comment (kurtamohler) [reply torchrl/envs/transforms/transforms.py:3106]: Oh right, `:num_repeats` will work, without the 0. I'll update it
